### DISCUSSION TITLE
How to use NextJS Link with Chakra UI Tooltip appropriate way

### DISCUSSION THREAD

**Discussion by zummon** [chakra-ui/chakra-ui#3856] [Q&A]
Hi I'm new here, I use **_Next_** JS **_Link_** and Chakra UI **_Button, Tooltip_** and import like so

    import NextLink from 'next/link'
    import { Button, Tooltip } from '@chakra-ui/react'

The code I'm trying to do is using **_Tooltip_** to show when I hover the **_Button_** to go to home page

    <NextLink href="/">
      <Tooltip label={t["to-home"]}>
        <Button size="lg" variant="link">
          zummon
        </Button>
      </Tooltip>
    </NextLink>

When I click it, it doesn't do anything (doesn't route to home page), so I try this to fix

      <Tooltip label={t["to-home"]}>
    <NextLink href="/">
        <Button size="lg" variant="link">
          zummon
        </Button>
    </NextLink>
      </Tooltip>

But even worse, error appears 

     Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?

     Check the render method of `Tooltip`.

**Are there appropriate ways to fix or work around, for this situation?**
**Thank you in advance**

**Comment by KristianRykkje** [comment]
This worked for me: 
```
<NextLink href="/">
 <Button size="lg" variant="link">
  <Tooltip label={t["to-home"]}>zummon</Tooltip>
 </Button>
</NextLink>
```

**Comment by zummon** [comment]
I thought it would be able to hover only the text, **not the whole button**. As variant="link" anyways, this works fine for me. Thank you @KristianRykkje

**Comment by payapula** [comment]
There are some challenges involved with what we are trying to achieve. We want to have the tooltip to show up when we hover over the whole button. 

> I thought it would be able to hover only the text, not the whole button

In order for tooltip to show up when we hover over the whole button, it must be the parent to the the button component and it must be child to the next/link. 

```jsx
<NextLink>
   <ToolTip>
      <Button>
        About
      </Button>
    </Tooltip>
</NextLink>
```


So, for this structure to work, we need 

- NextLink to `passHref` down to Tooltip, which should forward that to Button - [NextJs Docs](https://nextjs.org/docs/api-reference/next/link#if-the-child-is-a-function-component)

This code would do:

```jsx
import * as React from "react";
import NextLink from "next/link";
import { Button, Tooltip } from "@chakra-ui/react";

// the `props` will contain onClick, href and onMouseEnter from next/link
// ref needs to be passed to DOM element
const ToolTipButton = React.forwardRef(function ChakraLinkButton(props, ref) {
  return (
    <Tooltip label="to-about">
      <Button ref={ref} size="lg" {...props}>
        Go Home
      </Button>
    </Tooltip>
  );
});

export default function AboutPage() {
  return (
    <div>
      <NextLink href="/" passHref>
        <ToolTipButton />
      </NextLink>
    </div>
  );
}

```


However, to be **semantically correct**, `<button>` should not have a `href` attribute. So, we need to render `<a>` and not `<button>` 

We can style `Link` from Chakra to make it look like button with the help of `useStyleConfig` hook or we can also style it manually.

We can achieve these by the following code:

```jsx
import * as React from "react";
import NextLink from "next/link";
import { Link as ChakraLink, useStyleConfig, Tooltip } from "@chakra-ui/react";

// the `props` will contain onClick, href and onMouseEnter from next/link
// ref needs to be passed to DOM element
const ToolTipButton = React.forwardRef(function ChakraLinkButton(props, ref) {
  // More info on useStyleConfig here -
  // https://chakra-ui.com/docs/theming/component-style#usestyleconfig-api
  const styles = useStyleConfig("Button");
  return (
    <Tooltip label="to-about">
      <ChakraLink
        {...styles} // To make <a> look like chakra button
        display="block" // default <a> would be inline
        ref={ref}
        {...props}
      >
        About
      </ChakraLink>
    </Tooltip>
  );
});

export default function IndexPage() {
  return (
    <div>
      <NextLink href="/about" passHref>
        <ToolTipButton border="1px" height="30px" w="90px" />
      </NextLink>
    </div>
  );
}
```

You can play around with this code in this [sandbox](https://codesandbox.io/s/forked-nextjs-with-chakra-qdmn9?file=/pages/index.js). In this sandbox you would find both of the above approaches, I would prefer the second approach with Link rather than Button because it is valid HTML.
